File: crates/harness/tests/fixtures/prime-rpc.py
#!/usr/bin/env python3
"""A small Prime RPC peer for the harness's public journey test."""

import json
import pathlib
import sys


root = pathlib.Path(__file__).parent
with (root / "prime-args.log").open("a") as log:
    log.write(json.dumps(sys.argv[1:]) + "\n")


def send(value):
    print(json.dumps(value), flush=True)


session_file = str(root / "session.jsonl")
if "--resume" in sys.argv:
    session_file = sys.argv[sys.argv.index("--resume") + 1]

pending_prompt_id = None
steer_scenario = False
for line in sys.stdin:
    request = json.loads(line)
    kind = request["type"]
    request_id = request.get("id")
    if kind == "prompt":
        if request["message"] == "scenario:agent-goal":
            send({"type": "response", "id": request_id, "command": "prompt", "success": True})
            send({"type": "agent_start"})
            send({"type": "goal_update", "goal": {"status": "active", "objective": "Agent-created goal"}})
            send({"type": "turn_end", "message": {"role": "assistant"}, "toolResults": []})
            send({"type": "agent_end"})
            continue
        if request["message"] == "/goal status":
            send({"type": "response", "id": request_id, "command": "prompt", "success": True})
            send({"type": "session_action_update", "actions": {"active": {"kind": "session_command", "label": "/goal status"}}})
            send({"type": "goal_update", "goal": {"status": "active", "objective": "Finish work"}})
            send({"type": "message_end", "message": {"role": "custom", "customType": "session_slash_command_result", "content": "Goal active: Finish work", "display": True, "details": {"command": {"text": "/goal status"}, "success": True}}})
            send({"type": "session_action_update", "actions": {"queuedCount": 0}})
            continue
        if request["message"].startswith("/goal "):
            with (root / "prime-goal-actions.log").open("a") as log:
                log.write(request["message"] + "\n")
            if request["message"] == "/goal resume":
                send({"type": "response", "id": request_id, "command": "prompt", "success": False,
                      "error": "goal cannot resume"})
                continue
            send({"type": "response", "id": request_id, "command": "prompt", "success": True})
            send({"type": "session_action_update", "actions": {"active": {"kind": "session_command", "label": request["message"]}}})
            goal = None if request["message"] == "/goal clear" else {"status": "paused", "objective": "Finish work"}
            send({"type": "goal_update", "goal": goal})
            send({"type": "session_action_update", "actions": {"queuedCount": 0}})
            continue
        if request.get("streamingBehavior") == "steer":
            assert steer_scenario and request["message"] == "Redirect while child runs"
            send({"type": "response", "id": request_id, "command": "prompt", "success": True})
            send({"type": "observed_session_event", "activeSessionId": "active-child-1",
                  "event": {"type": "message_update", "assistantMessageEvent": {"type": "text_delta", "delta": " and finished"}}})
            send({"type": "rlm_child_update", "child": {"id": "child-1", "label": "child task", "status": "done", "activeSessionId": "active-child-1"}})
            send({"type": "observed_session_event", "activeSessionId": "active-child-1", "event": {"type": "agent_end"}})
            send({"type": "turn_end", "message": {"role": "assistant"}, "toolResults": []})
            send({"type": "agent_end"})
            continue
        steer_scenario = request["message"] == "parent steer scenario"
        pending_prompt_id = request_id
        send({"type": "extension_ui_request", "id": "confirm-1", "method": "confirm",
              "title": "Extension", "message": "Continue?"})
        continue
    if kind == "extension_ui_response":
        assert request["id"] == "confirm-1" and request["confirmed"] is True
        send({"type": "response", "id": pending_prompt_id, "command": "prompt", "success": True})
        send({"type": "agent_start"})
        send({"type": "turn_start"})
        send({"type": "session_action_update", "actions": {"queuedCount": 1}})
        send({"type": "compaction_start", "reason": "threshold"})
        send({"type": "compaction_end", "reason": "threshold", "result": {"summary": "saved"}})
        send({"type": "auto_retry_start", "attempt": 1, "maxAttempts": 3})
        send({"type": "auto_retry_end", "attempt": 1, "success": True})
        send({"type": "extension_ui_request", "id": "status-1", "method": "setStatus", "statusKey": "work", "statusText": "Running"})
        send({"type": "future_prime_event", "extra": {"value": 42}})
        send({"type": "message_update", "assistantMessageEvent": {"type": "text_delta", "delta": "parent live"}})
        send({"type": "tool_execution_update", "toolCallId": "tool-1", "partialResult": {"content": [{"type": "text", "text": "partial"}]}})
        send({"type": "rlm_child_update", "child": {"id": "child-1", "label": "child task", "status": "queued"}})
        send({"type": "rlm_child_update", "child": {"id": "child-1", "label": "child task", "status": "running", "activeSessionId": "active-child-1"}})
        continue
    if kind == "set_model":
        assert request["provider"] == "other" and request["modelId"] == "second"
    if kind == "get_state":
        data = {"sessionFile": session_file, "model": {"provider": "local", "id": "configured"},
                "thinkingLevel": "high", "isStreaming": False,
                "goal": {"status": "active", "objective": "Finish work"}}
    elif kind == "get_available_models":
        data = {"models": [
            {"provider": "local", "id": "configured", "name": "Configured", "reasoning": True,
             "thinkingLevelMap": {"minimal": None, "max": None}},
            {"provider": "other", "id": "second", "name": "Second", "reasoning": False},
        ]}
    elif kind == "get_commands":
        data = {"commands": [
            {"name": "skill:research", "source": "skill", "description": "Research",
             "sourceInfo": {"path": str(root / "SKILL.md")}},
            {"name": "extension-command", "source": "extension", "description": "Extension"},
        ]}
    elif kind == "observe":
        assert request["activeSessionId"] == "active-child-1"
        data = {"messages": [{"role": "user", "content": "child task"},
                             {"role": "assistant", "content": [{"type": "text", "text": "child history"}]}]}
    else:
        data = None

    send({"type": "response", "id": request_id, "command": kind, "success": True, "data": data})

    if kind == "observe":
        send({"type": "observed_session_event", "activeSessionId": "active-child-1",
              "event": {"type": "message_update", "assistantMessageEvent": {"type": "text_delta", "delta": "child live"}}})
        if steer_scenario:
            continue
        send({"type": "observed_session_event", "activeSessionId": "active-child-1",
              "event": {"type": "tool_execution_start", "toolCallId": "python-1", "toolName": "ipython", "args": {"code": "print(1)"}}})
        send({"type": "observed_session_event", "activeSessionId": "active-child-1",
              "event": {"type": "tool_execution_update", "toolCallId": "python-1", "partialResult": {"content": [{"type": "text", "text": "working"}]}}})
        send({"type": "observed_session_event", "activeSessionId": "active-child-1",
              "event": {"type": "tool_execution_end", "toolCallId": "python-1", "isError": False,
                        "result": {"content": [{"type": "text", "text": "1"}]}}})
        send({"type": "rlm_child_update", "child": {"id": "child-1", "label": "child task", "status": "done", "activeSessionId": "active-child-1"}})
        send({"type": "observed_session_event", "activeSessionId": "active-child-1", "event": {"type": "agent_end"}})
        send({"type": "turn_end", "message": {"role": "assistant"}, "toolResults": []})
        send({"type": "agent_end"})

File: crates/harness/tests/prime_rpc.rs
#![cfg(unix)]

use std::os::unix::fs::PermissionsExt;
use std::time::Duration;

use futures::StreamExt;
use tokio::sync::{mpsc, oneshot};
use zeron_harness::{CancellationToken, Harness, PrimeHarness, RunControls, SteerMessage};
use zeron_proto::{
    AgentEvent, DoneStatus, ReasoningLevel, RunRequest, SandboxLevel, ToolCall, UserInputAnswer,
};

#[tokio::test]
async fn session_command_finishes_without_an_agent_turn() {
    let dir = tempfile::tempdir().unwrap();
    let executable = dir.path().join("prime-agent");
    std::fs::write(&executable, include_str!("fixtures/prime-rpc.py")).unwrap();
    std::fs::set_permissions(&executable, std::fs::Permissions::from_mode(0o755)).unwrap();
    let harness = PrimeHarness::new().with_executable(&executable);
    let (steer, steering) = mpsc::channel(1);
    drop(steer);
    let controls = RunControls {
        steering,
        interrupt: CancellationToken::new(),
        request_input: Box::new(|_| {
            let (tx, rx) = oneshot::channel();
            let _ = tx.send(Vec::new());
            rx
        }),
    };
    let request = RunRequest {
        prompt: "/goal status".into(),
        harness: Some(zeron_proto::HarnessId::Prime),
        model: Some("default".into()),
        reasoning: None,
        model_options: Default::default(),
        cwd: dir.path().display().to_string(),
        sandbox: SandboxLevel::WorkspaceWrite,
        auto_approve: true,
        resume: None,
        attachments: Vec::new(),
        worktree: None,
    };
    let mut stream = harness.run(request, controls).await.unwrap();
    let events = tokio::time::timeout(Duration::from_secs(5), async {
        let mut events = Vec::new();
        while let Some(event) = stream.next().await {
            let event = event.unwrap();
            let done = matches!(event, AgentEvent::Done { .. });
            events.push(event);
            if done {
                break;
            }
        }
        events
    })
    .await
    .expect("Prime command did not finish");
    assert!(events.iter().any(|event| matches!(event, AgentEvent::TextDelta { text } if text == "Goal active: Finish work")));
    assert!(matches!(
        events.last(),
        Some(AgentEvent::Done {
            status: DoneStatus::Completed,
            ..
        })
    ));
}

#[tokio::test]
async fn local_catalog_and_persistent_child_stream_follow_prime_rpc() {
    let dir = tempfile::tempdir().unwrap();
    let executable = dir.path().join("prime-agent");
    std::fs::write(&executable, include_str!("fixtures/prime-rpc.py")).unwrap();
    std::fs::set_permissions(&executable, std::fs::Permissions::from_mode(0o755)).unwrap();
    let harness = PrimeHarness::new().with_executable(&executable);

    let models = harness.models().await.unwrap();
    assert_eq!(models.len(), 3);
    assert_eq!(models[0].id, "default");
    assert!(models[0].reasoning_levels.is_empty());
    assert_eq!(models[1].id, "local/configured");
    assert!(
        !models[1]
            .reasoning_levels
            .contains(&ReasoningLevel::Minimal)
    );
    assert!(!models[1].reasoning_levels.contains(&ReasoningLevel::Max));
    assert_eq!(models[2].id, "other/second");
    let skills = harness.skills(dir.path()).await.unwrap().unwrap();
    assert_eq!(skills.len(), 1);
    assert_eq!(skills[0].name, "research");
    let commands = harness.commands_for(dir.path()).await.unwrap();
    for name in ["compact", "refine", "goal", "autonomous"] {
        assert!(commands.iter().any(|command| command.name == name));
    }
    assert_eq!(commands.len(), 6);

    let session_file = dir.path().join("session.jsonl").display().to_string();
    for (resume, model, expected_model) in [
        (None, "default", "local/configured"),
        (Some(session_file.clone()), "other/second", "other/second"),
    ] {
        let expected_initial = resume.as_ref().map(|_| 12_000);
        let (_steer_tx, steering) = mpsc::channel(1);
        let controls = RunControls {
            steering,
            interrupt: CancellationToken::new(),
            request_input: Box::new(|questions| {
                let (tx, rx) = oneshot::channel();
                let _ = tx.send(vec![UserInputAnswer {
                    question_id: questions[0].id.clone(),
                    labels: vec!["Yes".into()],
                }]);
                rx
            }),
        };
        let request = RunRequest {
            prompt: "parent task".into(),
            harness: None,
            model: Some(model.into()),
            reasoning: None,
            model_options: Default::default(),
            cwd: dir.path().display().to_string(),
            sandbox: SandboxLevel::WorkspaceWrite,
            auto_approve: true,
            resume,
            attachments: Vec::new(),
            worktree: None,
        };
        let mut stream = harness.run(request, controls).await.unwrap();
        let events = tokio::time::timeout(Duration::from_secs(10), async {
            let mut events = Vec::new();
            while let Some(event) = stream.next().await {
                let event = event.unwrap();
                let parent_done = matches!(event, AgentEvent::Done { .. });
                events.push(event);
                if parent_done {
                    break;
                }
            }
            events
        })
        .await
        .unwrap();
        assert!(events.iter().any(|event| matches!(event,
            AgentEvent::SessionStarted { session_id, model, .. }
                if session_id == &session_file && model == expected_model)));
        let context: Vec<_> = events
            .iter()
            .filter_map(|event| match event {
                AgentEvent::ContextUsageSnapshot { tokens, window } => Some((*tokens, *window)),
                _ => None,
            })
            .collect();
        assert_eq!(
            context,
            vec![
                (expected_initial, Some(200_000)),
                (None, Some(200_000)),
                (Some(42_000), Some(200_000)),
            ]
        );
        assert!(events.iter().any(|event| matches!(event,
            AgentEvent::ToolCall { call: ToolCall::Unknown { name, .. }, .. }
                if name == "Agent: child task")));
        assert!(events.iter().any(|event| matches!(event,
            AgentEvent::Subagent { event, .. }
                if matches!(event.as_ref(), AgentEvent::TextDelta { text } if text == "child live"))));
        assert!(events.iter().any(|event| matches!(event,
            AgentEvent::Subagent { event, .. }
                if matches!(event.as_ref(), AgentEvent::ToolCall { call: ToolCall::Exec { command }, .. } if command == "print(1)"))));
        assert!(events.iter().any(|event| matches!(event,
            AgentEvent::Subagent { event, .. }
                if matches!(event.as_ref(), AgentEvent::Done { status: DoneStatus::Completed, .. }))));
        let native: Vec<_> = events
            .iter()
            .filter_map(|event| match event {
                AgentEvent::PrimeEvent { event } => Some(event),
                _ => None,
            })
            .collect();
        for kind in [
            "extension_ui_request",
            "agent_start",
            "turn_start",
            "session_action_update",
            "compaction_start",
            "compaction_end",
            "auto_retry_start",
            "auto_retry_end",
            "future_prime_event",
            "message_update",
            "tool_execution_update",
            "rlm_child_update",
            "observed_session_event",
            "turn_end",
            "agent_end",
        ] {
            assert!(
                native.iter().any(|event| event["type"] == kind),
                "missing native {kind}"
            );
        }
        assert!(native.iter().any(|event| event["type"] == "future_prime_event"
            && event["extra"]["value"] == 42));
        assert!(
            native
                .iter()
                .any(|event| event["type"] == "observed_session_event"
                    && event["event"]["type"] == "tool_execution_update"
                    && event["event"]["partialResult"]["content"][0]["text"] == "working")
        );
        assert!(matches!(
            events.last(),
            Some(AgentEvent::Done {
                status: DoneStatus::Completed,
                ..
            })
        ));
        drop(stream);
    }
    let args = std::fs::read_to_string(dir.path().join("prime-args.log")).unwrap();
    assert!(
        args.lines()
            .any(|line| line.contains("--resume") && line.contains(&session_file))
    );
}

#[tokio::test]
async fn native_steer_keeps_a_running_prime_child_alive() {
    let dir = tempfile::tempdir().unwrap();
    let executable = dir.path().join("prime-agent");
    std::fs::write(&executable, include_str!("fixtures/prime-rpc.py")).unwrap();
    std::fs::set_permissions(&executable, std::fs::Permissions::from_mode(0o755)).unwrap();
    let harness = PrimeHarness::new().with_executable(&executable);
    let (steer_tx, steering) = mpsc::channel(1);
    let controls = RunControls {
        steering,
        interrupt: CancellationToken::new(),
        request_input: Box::new(|questions| {
            let (tx, rx) = oneshot::channel();
            let _ = tx.send(vec![UserInputAnswer {
                question_id: questions[0].id.clone(),
                labels: vec!["Yes".into()],
            }]);
            rx
        }),
    };
    let request = RunRequest {
        prompt: "parent steer scenario".into(),
        harness: Some(zeron_proto::HarnessId::Prime),
        model: Some("default".into()),
        reasoning: None,
        model_options: Default::default(),
        cwd: dir.path().display().to_string(),
        sandbox: SandboxLevel::WorkspaceWrite,
        auto_approve: true,
        resume: None,
        attachments: Vec::new(),
        worktree: None,
    };
    let mut stream = harness.run(request, controls).await.unwrap();
    let events = tokio::time::timeout(Duration::from_secs(10), async {
        let mut events = Vec::new();
        while let Some(event) = stream.next().await {
            let event = event.unwrap();
            let child_started = matches!(&event, AgentEvent::Subagent { event, .. }
                if matches!(event.as_ref(), AgentEvent::TextDelta { text } if text == "child live"));
            let done = matches!(event, AgentEvent::Done { .. });
            events.push(event);
            if child_started {
                steer_tx
                    .send(SteerMessage {
                        prompt: "Redirect while child runs".into(),
                        message_id: Some("steered-user".into()),
                    })
                    .await
                    .unwrap();
            }
            if done {
                break;
            }
        }
        events
    })
    .await
    .expect("Prime child and parent finish after steer");
    assert!(
        events
            .iter()
            .any(|event| matches!(event, AgentEvent::Subagent { event, .. }
        if matches!(event.as_ref(), AgentEvent::TextDelta { text } if text == " and finished")))
    );
    assert!(
        events
            .iter()
            .any(|event| matches!(event, AgentEvent::Subagent { event, .. }
        if matches!(event.as_ref(), AgentEvent::Done { status: DoneStatus::Completed, .. })))
    );
    assert!(matches!(
        events.last(),
        Some(AgentEvent::Done {
            status: DoneStatus::Completed,
            ..
        })
    ));
    assert!(!events.iter().any(|event| matches!(
        event,
        AgentEvent::Done {
            status: DoneStatus::Interrupted,
            ..
        }
    )));
}
